### PR TITLE
support texture binding when bindless is not available

### PR DESCRIPTION
**Pull request by ruancomelli**
## Summary by CodeRabbit

- **New Features**
	- Introduced command-line argument handling for enhanced application configurability.
	- Added support for controlling bindless texture functionality.
	- Improved texture management with dynamic shader assignment based on texture state.

- **Bug Fixes**
	- Enhanced error handling for texture initialization and rendering processes to prevent null reference issues.

- **Documentation**
	- Updated comments and debug prints to clarify new command-line options and texture management logic.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is a benchmark review for experiment `multiple_mermaid_diagram_types`.
Run ID: `multiple_mermaid_diagram_types/benchmark_2024-09-26T14-52-30_v1-23-0-20-g8a080be57`.

This pull request was cloned from `https://github.com/btipling/foundations/pull/33`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_
<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
    # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
    # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35
  # Including database changes to evaluate entity-relationship diagrams
- https://github.com/sourcery-ai/core/pull/4579
  # TODO: find some with Alembic

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/370
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/371
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/372
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/3
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/11
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/12
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/12
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/11
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/3
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/11
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/21
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/11
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/22
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/6
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/11
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/12
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/6
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/6
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/11
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/12
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/6
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/6
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/6
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/4
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/12
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/13
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/13
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/12
- dataset_pull_request: https://github.com/4DNucleome/PartSeg/pull/1183
  review_pull_request: https://github.com/sourcery-ai-experiments/PartSeg/pull/6
- dataset_pull_request: https://github.com/ably/ably-cocoa/pull/1973
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-cocoa/pull/5
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/54
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/9
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/35
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/10
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4579
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/373

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request implements support for texture binding when bindless textures are not available. It introduces a new command-line argument to disable bindless textures, modifies the texture initialization and binding process, and updates the shader selection based on the texture state. The changes primarily affect the Texture module and various scene implementations.

### Sequence Diagram

```mermaid
sequenceDiagram
    participant App
    participant Args
    participant Texture
    participant Scene
    participant Shader

    App->>Args: Parse command-line arguments
    Args-->>App: Return disable_bindless flag
    App->>Scene: Initialize with args
    Scene->>Texture: init(disable_bindless)
    Texture-->>Scene: Return Texture object
    Scene->>Texture: setup(image, program, uniform_name)
    Texture->>Texture: Check bindless support
    alt Bindless supported and not disabled
        Texture->>Texture: Create bindless handle
    else Bindless not supported or disabled
        Texture->>Texture: Prepare for non-bindless use
    end
    Texture-->>Scene: Texture ready
    Scene->>Shader: Create with dynamic fragment shader type
    Shader-->>Scene: Return Shader object
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Introduced command-line argument handling for disabling bindless textures | <ul><li>Added a new Args struct to handle command-line arguments</li><li>Implemented parsing of the '--disable_bindless' flag</li><li>Added help text for the new command-line option</li></ul> | `src/foundations/Args.zig`<br/>`src/foundations/app.zig`<br/>`src/foundations/scenes/scenes.zig` |
| Modified Texture initialization and binding process | <ul><li>Added a 'disable_bindless' flag to the Texture struct</li><li>Split texture initialization into two steps: init and setup</li><li>Implemented fallback to non-bindless textures when bindless is not supported or disabled</li><li>Added support for creating a default magenta texture when no image is provided</li></ul> | `src/foundations/rhi/Texture.zig` |
| Updated shader selection based on texture state | <ul><li>Added a new function 'frag_shader' to determine the appropriate fragment shader type</li><li>Modified scene implementations to use the new texture initialization process</li><li>Updated shader creation to use the dynamically determined fragment shader type</li></ul> | `src/foundations/rhi/Texture.zig`<br/>`src/foundations/scenes/cgpoc/chapter5/textured_pyramid/TexturedPyramid.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/textured_torus/TexturedTorus.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/earth/Earth.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/shuttle/Shuttle.zig` |
| Improved error handling and fallback mechanisms | <ul><li>Removed the 'BindlessNotSupported' error and replaced it with automatic fallback</li><li>Added null checks and error handling for texture initialization in scene implementations</li><li>Implemented graceful fallback to non-bindless textures when initialization fails</li></ul> | `src/foundations/rhi/Texture.zig`<br/>`src/foundations/scenes/cgpoc/chapter5/textured_pyramid/TexturedPyramid.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/textured_torus/TexturedTorus.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/earth/Earth.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/shuttle/Shuttle.zig` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourceryai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/foundations/rhi/Texture.zig [hunk 1]
```diff
@@ -4,45 +4,68 @@ handle: c.GLuint64 = 0,
...
+    };
+
     if (c.glfwExtensionSupported("GL_ARB_bindless_texture") != 1 or c.glfwExtensionSupported("GL_ARB_bindless_texture") != 1) {
-        return TextureError.BindlessNotSupported;
+        t.disable_bindless = true;
     }
+    return t;
+}
 
+pub fn setup(self: *Texture, image: ?*assets.Image, program: u32, uniform_name: []const u8) TextureError!void {
```
Comment (SourceryAI) [inline src/foundations/rhi/Texture.zig:31]: **suggestion (bug_risk):** Consider more robust error handling in texture setup

The current implementation might silently fail in some error cases. Consider propagating errors or logging them for better debugging and reliability.

File: src/foundations/scenes/cgpoc/chapter5/textured_pyramid/TexturedPyramid.zig [hunk 2]
```diff
@@ -88,11 +88,13 @@ pub fn updateCamera(_: *TexturedPyramid) void {}
 
 pub fn renderParallepiped(self: *TexturedPyramid) void {
     const prog = rhi.createProgram();
+    self.brick_texture = rhi.Texture.init(self.ctx.args.disable_bindless) catch null;
```
Comment (SourceryAI) [inline src/foundations/scenes/cgpoc/chapter5/textured_pyramid/TexturedPyramid.zig:91]: **suggestion (performance):** Evaluate performance impact of disabling bindless textures

While the flexibility is good, consider documenting or measuring the performance difference between bindless and non-bindless textures to guide users on when to use this option.

File: src/foundations/rhi/Texture.zig [hunk 3]
```diff
@@ -4,45 +4,68 @@ handle: c.GLuint64 = 0,
...
-        0,
-        0,
-        0,
-        @intCast(image.width),
-        @intCast(image.height),
-        c.GL_RGBA,
-        c.GL_UNSIGNED_BYTE,
-        image.stb_data.ptr,
-    );
+    if (image) |img| {
```
Comment (SourceryAI) [inline src/foundations/rhi/Texture.zig:38]: **suggestion:** Consider handling texture creation failure

The code creates a magenta texture when no image is provided, but doesn't handle the case where texture creation fails. Consider adding error handling for this scenario.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 3 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/95740ae1-c4bc-4525-bea1-6c0c78323524?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**